### PR TITLE
Add option to set :noindex:

### LINKED ISSUE #130: Incompatible with sphinx>4 (submodules of submodules)
Not sure what is happening but we need to make sure it is not a bug here before reporting it upstream to `sphinx`. This is blocking astropy/astropy#11763 .

@larrybradley narrowed down the issue in https://github.com/astropy/astropy/issues/11723#issuecomment-838733176

<hr/>

I think this is a real issue.  I'm getting this same error in `photutils`.  It appears to be due to submodules of submodules:

```
mod1
    __init__.py  (from mod2 import *)
    mod2
       __init__.py (from a import *)
       a.py (defines myfunc, and __all__ = ['myfunc'])
```

In this case, Sphinx is now defining duplicate references as:
`package.mod1.myfunc` and `package.mod1.mod2.myfunc` and raising the
"duplicate object descriptions".

This started with Sphinx 4 release.  Not sure if this was an unintended bug there or if something needs to be fixed in `astropy-sphinx` for compatiblity with Sphinx 4.

### PR DESCRIPTION
**Pull request by saimn**
Fix #130.

TODO: see if I can add a test for this.

### CODE REVIEW

**Comment by saimn** [comment]
Wasn't easy but I managed to add a test :)
Who wants to review ?

**Comment by saimn** [comment]
Oh, of course tests fail... maybe for old sphinx versions though

**Comment by saimn** [comment]
All green now :)

**Comment by pllim** [comment]
Ooooo... how do we use this in `astropy`? Does it just magically work?

**Comment by saimn** [comment]
No magic :D, we must add `:noindex:` to the duplicated directives, and this PR just propagates `:noindex:` to autodoc.

**Comment by pllim** [comment]
@saimn , is there an `astropy` PR that uses this feature? Would be nice to see if it really works before we merge.

File: sphinx_automodapi/tests/test_cases.py [hunk 1]
```diff
@@ -115,3 +115,28 @@ def test_run_full_case(tmpdir, case_dir, parallel):
             with open(path_reference, encoding='utf8') as f:
                 reference = f.read()
             assert actual.strip() == reference.strip()
+
+
+def test_duplicated_warning(tmpdir):
```
Comment (astrofrog) [inline sphinx_automodapi/tests/test_cases.py:120]: Just to check, why can't this use the ``test_run_full_case`` machinery above?

**Comment by saimn** [comment]
> is there an astropy PR that uses this feature? Would be nice to see if it really works before we merge.

@pllim - no, I could do that indeed :+1:

**Comment by pllim** [comment]
@saimn , any chance we can push this forward? Did you say there is something that not quite working still using this with `astropy`? 

@WilliamJamieson is unable to build `astropy` docs on Python 3.10 because of the Sphinx maxversion pin.

**Comment by saimn** [comment]
Within another round of testing/debugging on Astropy I'm pretty convinced that the changes here are correct and are doing what is supposed. Adding `:noindex:` should help fixing duplicate warning for simple cases such as https://github.com/maxnoe/sphinx-automodapi-duplicated-warning, and probably photutils. But for astropy it's not enough because of the complicated structure of the docs and code, using `:noindex:` there then creates some other warnings about missing references.

**Comment by pllim** [comment]
Is it time to make astropy less complicated?! We cannot pin Sphinx forever.

**Comment by Cadair** [comment]
> Is it time to make astropy less complicated?!

I have been working on this for years :stuck_out_tongue:

**Review by Cadair — APPROVED** [review]
:shipit:

**Review by pllim — APPROVED** [review]
Just some nitpicks. If you want to address them, great. If not, we can defer. Though that one typo should probably be fixed anyway before merge.

Let's ship this indeed and then fix up astropy! 🚀

File: sphinx_automodapi/tests/duplicated_warning/docs/conf.py [hunk 2]
```diff
@@ -0,0 +1,11 @@
+project = 'duplicated'
+copyright = '2022, Maximilian Nöthe'
```
[the remark this answers is not in the document]
Comment (maxnoe) [reply sphinx_automodapi/tests/duplicated_warning/docs/conf.py:2]: sure

[the remark this answers is not in the document]
Comment (maxnoe) [reply sphinx_automodapi/tests/duplicated_warning/docs/conf.py:2]: Though I am Maximilian Linhoff now ;)

**Comment by Cadair** [comment]
Even though astropy core doesn't need this, we might as well merge it?

**Comment by pllim** [comment]
Gah, CI got disabled. I have to close/reopen.

**Comment by pllim** [comment]
Merged! I'll let someone else do the release. Thanks, all!

**Pull request merged**